src/v2/kuadrant/pipeline/tasks/auth.rs

```diff
@@ -0,0 +1,238 @@
+            predicates,
+            dependencies,
+            is_blocking,
+        }
+    }
+}
+
+impl Task for AuthTask {
+    fn prepare(&self, ctx: &mut ReqRespCtx) -> TaskOutcome {
+        let mut has_pending = false;
+        let mut has_error = false;
+
+        macro_rules! touch {
+            ($path:expr, $type:ty) => {
+                match ctx.get_attribute::<$type>($path) {
+                    Ok(AttributeState::Available(_)) => {}
+                    Ok(AttributeState::Pending) => has_pending = true,
+                    Err(_) => has_error = true,
+                }
+            };
+        }
+
+        // these are required to build a CheckRequest
+        touch!("request.headers", Headers);
+        touch!("request.host", String);
+        touch!("request.method", String);
+        touch!("request.scheme", String);
+        touch!("request.path", String);
+        touch!("request.protocol", String);
+        touch!("request.time", DateTime<FixedOffset>);
+        touch!("destination.address", String);
+        touch!("destination.port", i64);
+        touch!("source.address", String);
+        touch!("source.port", i64);
+
+        // eval request_data early
+        let _ = ctx.eval_request_data();
+
+        // in this case they should all be available
+        if has_pending || has_error {
+            TaskOutcome::Failed
+        } else {
+            TaskOutcome::Done
+        }
+    }
+
+    fn id(&self) -> Option<String> {
+        Some(self.task_id.clone())
+    }
+
+    fn dependencies(&self) -> &[String] {
+        &self.dependencies
+    }
+
+    fn apply(self: Box<Self>, ctx: &mut ReqRespCtx) -> TaskOutcome {
+        match self.predicates.apply(ctx) {
+            Ok(AttributeState::Pending) => return TaskOutcome::Requeued(vec![self]),
+            Ok(AttributeState::Available(false)) => return TaskOutcome::Done,
+            Ok(AttributeState::Available(true)) => {}
+            Err(_e) => {
+                return TaskOutcome::Failed;
+            }
+        }
+
+        let token_id = match self.service.dispatch_auth(ctx, &self.scope) {
+            Ok(id) => id,
+            Err(_e) => {
+                return TaskOutcome::Failed;
+            }
+        };
+
+        let service = self.service.clone();
+
+        TaskOutcome::Deferred {
+            token_id,
+            pending: PendingTask {
+                task_id: Some(self.task_id),
+                is_blocking: self.is_blocking,
+                process_response: Box::new(move |ctx, status_code, response_size| {
+                    if status_code != proxy_wasm::types::Status::Ok as u32 {
+                        // todo(refactor): failure case
+                        return TaskOutcome::Failed;
+                    }
+
+                    match service.get_response(ctx, response_size) {
+                        Ok(parsed) => process_auth_response(parsed),
+                        Err(_e) => TaskOutcome::Failed,
+                    }
```

Comment:
The `status_code` gets passed here so we can decide what to do with it; and then defers to the service to grab the message